internal/recover/recover.go

```diff
@@ -0,0 +1,339 @@
+func ReadTrustStore(dir string) (*trust.Remotes, error) {
+	remotes := &trust.Remotes{}
+	err := remotes.Load(dir)
+
+	return remotes, err
+}
+
+// ValidateMemberChanges compares two arrays of members to ensure:
+// - Their lengths are the same.
+// - Members with the same name also use the same ID and address.
+func ValidateMemberChanges(oldMembers []cluster.LocalMember, newMembers []cluster.LocalMember) error {
+	if len(newMembers) != len(oldMembers) {
+		return fmt.Errorf("members cannot be added or removed")
+	}
+
+	for _, newMember := range newMembers {
+		memberValid := false
+		for _, oldMember := range oldMembers {
+			// FIXME: Allow changing member addresses as part of cluster recovery
+			membersMatch := newMember.DqliteID == oldMember.DqliteID &&
+				newMember.Name == oldMember.Name &&
+				newMember.Address == oldMember.Address
+
+			if membersMatch {
+				memberValid = true
+				break
+			}
+		}
+
+		if !memberValid {
+			return fmt.Errorf("ID or address changed for member %s", newMember.Name)
+		}
+	}
+
+	return nil
+}
+
+// CreateRecoveryTarball writes a tarball of filesystem.DatabaseDir to
+// filesystem.StateDir.
+// go-dqlite's info.yaml is excluded from the tarball.
+func CreateRecoveryTarball(filesystem *sys.OS) error {
+	dbFS := os.DirFS(filesystem.DatabaseDir)
+	dbFiles, err := fs.Glob(dbFS, "*")
+	if err != nil {
+		return fmt.Errorf("%w", err)
+	}
+
+	tarballPath := path.Join(filesystem.StateDir, "recovery_db.tar.gz")
+
+	// info.yaml is used by go-dqlite to keep track of the current cluster member's
+	// ID and address. We shouldn't replicate the recovery member's info.yaml
+	// to all other members, so exclude it from the tarball:
+	for indx, filename := range dbFiles {
+		if filename == "info.yaml" {
+			newlen := len(dbFiles) - 1
+			dbFiles[indx] = dbFiles[newlen]
+			dbFiles = dbFiles[:newlen]
+			break
+		}
+	}
+
+	return createTarball(tarballPath, filesystem.DatabaseDir, dbFiles)
+}
+
+// MaybeUnpackRecoveryTarball checks for the presence of a recovery tarball in
+// fiesystem.StateDir. If it exists, unpack it into a temporary directory,
+// ensure that it is a valid microcluster recovery tarball, and replace the
+// existing filesystem.DatabaseDir.
+func MaybeUnpackRecoveryTarball(filesystem *sys.OS) error {
+	tarballPath := path.Join(filesystem.StateDir, "recovery_db.tar.gz")
+	unpackDir := path.Join(filesystem.StateDir, "recovery_db")
+
+	// Determine if the recovery tarball exists
+	if _, err := os.Stat(tarballPath); errors.Is(err, os.ErrNotExist) {
+		return nil
+	}
+
+	logger.Warn("Recovery tarball located; attempting DB recovery", logger.Ctx{"tarball": tarballPath})
+
+	err := unpackTarball(tarballPath, unpackDir)
+	if err != nil {
+		return err
+	}
+
+	// TODO Is this a reasonable sanity check or is a parseable cluster.yaml
+	// containing the local node sufficient?
+	metadataPath := path.Join(unpackDir, "metadata1")
+	if _, err := os.Stat(metadataPath); errors.Is(err, os.ErrNotExist) {
```

Comment:
Is this file always present, both when the daemon is stopped cleanly and uncleanly?

Reply:
Unclear. I initially added this bit before realizing that checking if the local node is in the incoming `cluster.yaml` was a much more robust check. Going to drop this bit.